File: Data Structures & Algorithms/combination-target-sum/submission-3.cpp
class Solution {
    vector<vector<int>> ret;
private:
    void dfs(vector<int>& nums, int sum, int target, int start, vector<int>& res){
        if (target == sum){
            ret.push_back(res);
            return;
        }
        for (int i = start ; i < nums.size() ; i++){
            if (sum + nums[i] > target){
                break;
            }
            res.push_back(nums[i]);
            dfs(nums, sum + nums[i], target, i, res);
            res.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<int> res;
        dfs(nums, 0, target, 0, res);
        return ret;
    }
};

File: Data Structures & Algorithms/search-in-rotated-sorted-array-ii/submission-3.cs
public class Solution {
    public bool Search(int[] nums, int target) {
        int l = 0;
        int r = nums.Length - 1;

        while(l <= r){
            int m = l + (r - l)/2;
            if (target == nums[m]) return true;

            //left portion
            if (nums[l] < nums[m]){
                if (target < nums[m] && target >= nums[l]) r = m - 1;
                else l = m + 1;
            }else if (nums[l] > nums[m]){
                if (target > nums[m] && target <= nums[r]) l = m + 1;
                else r = m - 1;
            }else l++;
        }return false;
    }
}